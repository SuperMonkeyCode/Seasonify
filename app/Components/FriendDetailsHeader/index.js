/* @flow */
import { connect } from 'react-redux';

import { editActions, editSelectors } from 'State/editState';
import { friendActions, friendSelectors } from 'State/friendState';
import { uiActions } from 'State/uiState';

import FriendDetailsHeader from './FriendDetailsHeader';

export default connect(
  (state, ownProps) => ({
    isEditing: editSelectors.isEditing(state),
    editedObject: editSelectors.editedObject(state),
    friend: friendSelectors.getFriend(state, ownProps.friendId),
  }),
  {
    friendDelete: friendActions.friendDelete,
    friendUpdate: friendActions.friendUpdate,
    messageToast: uiActions.messageToast,
    editStart: editActions.editStart,
    editEnd: editActions.editEnd,
  },
)(FriendDetailsHeader);
