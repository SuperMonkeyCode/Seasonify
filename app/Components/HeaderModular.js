/* @flow */
import React from 'react';
import { Keyboard } from 'react-native';
import { Header, Title, Left, Right, Body, Button, Icon, Text } from 'native-base';
import type { HeaderButton } from 'Types';

export default function HeaderModular({
  title,
  cancelButton,
  actionButtons = [],
}: {
  title: string,
  cancelButton: HeaderButton,
  actionButtons: HeaderButton[],
}) {
  const cancelAction = () => {
    cancelButton.action();
    Keyboard.dismiss();
  };

  const renderTextOrIcon = (button: HeaderButton) => {
    if (button.icon) return <Icon name={button.icon} />;
    if (button.text) return <Text>{button.text}</Text>;
    return null;
  };

  const renderButton = () =>
    actionButtons.map((button, index) => {
      if (
        button.visibleIf ||
        (typeof button.visibleIf === 'undefined' && !button.hideByDefault)
      ) {
        return (
          /* eslint-disable react/no-array-index-key */
          <Button key={index} transparent onPress={button.action}>
            {renderTextOrIcon(button)}
          </Button>
        );
      }
      return null;
    });

  return (
    <Header>
      <Left>
        <Button transparent onPress={cancelAction}>
          {renderTextOrIcon(cancelButton)}
        </Button>
      </Left>
      <Body>
        <Title>{title}</Title>
      </Body>
      <Right>
        {renderButton()}
      </Right>
    </Header>
  );
}
