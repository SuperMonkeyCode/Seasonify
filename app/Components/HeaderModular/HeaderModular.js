/* @flow */
import { Header, Title, Left, Right, Body, Button, Icon, Text } from 'native-base';
import { Keyboard } from 'react-native';
import { compose, pure, withHandlers } from 'recompose';
import React from 'react';

type HeaderIconButton = {
  icon: string,
  action: Function,
  visibleIf?: boolean,
  hideByDefault?: boolean,
};
type HeaderTextButton = {
  text: string,
  action: Function,
  visibleIf?: boolean,
  hideByDefault?: boolean,
};
export type HeaderButton = HeaderIconButton | HeaderTextButton;

type Props = {
  /* parent */
  title: string,
  cancelButton: HeaderButton,
  actionButtons?: HeaderButton[],
  /* connect */
  /* HOC */
  handleCancel: Function,
};

const enhance = compose(
  withHandlers({
    handleCancel: ({ cancelButton }: Props) => () => {
      cancelButton.action();
      Keyboard.dismiss();
    },
  }),
  pure,
);

function HeaderModular({ title, cancelButton, actionButtons = [], handleCancel }: Props) {
  const renderTextOrIcon = (button: HeaderButton) => {
    if (button.icon) return <Icon name={button.icon} />;
    if (button.text) {
      return (
        <Text>
          {button.text}
        </Text>
      );
    }
    return null;
  };

  const renderButton = () =>
    actionButtons.map((button, index) => {
      if (
        button.visibleIf ||
        (typeof button.visibleIf === 'undefined' && !button.hideByDefault)
      ) {
        return (
          /* eslint-disable react/no-array-index-key */
          <Button key={index} transparent onPress={button.action}>
            {renderTextOrIcon(button)}
          </Button>
        );
      }
      return null;
    });

  return (
    <Header>
      <Left>
        <Button transparent onPress={handleCancel}>
          {renderTextOrIcon(cancelButton)}
        </Button>
      </Left>
      <Body>
        <Title>
          {title}
        </Title>
      </Body>
      <Right>
        {renderButton()}
      </Right>
    </Header>
  );
}

export default enhance(HeaderModular);
