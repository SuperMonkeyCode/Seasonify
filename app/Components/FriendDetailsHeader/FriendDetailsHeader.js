/* @flow */
import { Keyboard } from 'react-native';
import { compose, pure, withHandlers } from 'recompose';
import React from 'react';

import type { Friend } from 'Types';
import HeaderModular from 'Components/HeaderModular';

type Props = {
  /* parent */
  navigation: Object,
  friendId: string,
  /* connect */
  isEditing: boolean,
  editedObject: Object,
  friend: Friend,
  friendDelete: Function,
  friendUpdate: Function,
  messageToast: Function,
  editStart: Function,
  editEnd: Function,
  /* HOC */
  handleExit: Function,
  handleEdit: Function,
  handleDone: Function,
  handleDelete: Function,
};

const enhance = compose(
  pure,
  withHandlers({
    handleExit: ({ navigation, editEnd }: Props) => () => {
      navigation.goBack();
      Keyboard.dismiss();
      editEnd();
    },
    handleEdit: ({ editStart }: Props) => () => {
      editStart();
    },
    handleDone: ({
      editedObject,
      friendUpdate,
      friend,
      messageToast,
      editEnd,
    }: Props) => () => {
      Keyboard.dismiss();
      if (editedObject.id) {
        friendUpdate(editedObject);
        messageToast(
          'success',
          `“${friend.name}” edited${editedObject.name
            ? ` to “${editedObject.name}”`
            : ''}`,
        );
      }
      editEnd();
    },
    handleDelete: ({ navigation, friend, editEnd, friendDelete }: Props) => () => {
      navigation.goBack();
      Keyboard.dismiss();
      editEnd();
      friendDelete(friend.id);
    },
  }),
);

function FriendDetailsHeader({
  isEditing,
  friend,
  handleExit,
  handleEdit,
  handleDone,
  handleDelete,
}: Props) {
  return (
    <HeaderModular
      title={friend.name}
      cancelButton={{ icon: 'arrow-back', action: handleExit }}
      actionButtons={[
        {
          visibleIf: !isEditing,
          icon: 'create',
          action: handleEdit,
        },
        {
          visibleIf: isEditing,
          hideByDefault: true,
          icon: 'checkmark',
          action: handleDone,
        },
        { icon: 'trash', action: handleDelete },
      ]}
    />
  );
}

export default enhance(FriendDetailsHeader);
