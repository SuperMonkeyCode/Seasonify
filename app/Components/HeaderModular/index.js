import HeaderModular from './HeaderModular';

export default HeaderModular;
