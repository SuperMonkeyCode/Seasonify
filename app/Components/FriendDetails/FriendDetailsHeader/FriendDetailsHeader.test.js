import React from 'react';

import HeaderModular from 'Components/HeaderModular';

import FriendDetailsHeader from './FriendDetailsHeader';

function setup(specificProps = {}) {
  const props = {
    friendId: 'abc123',
    navigation: {
      navigate: jest.fn(),
      goBack: jest.fn(),
      state: { params: { friendId: 'abc123' } },
    },
    isEditing: true,
    editedObject: { id: 'abc123', name: 'Edited name' },
    friend: {
      id: 'abc123',
      name: 'Some friend',
    },
    friendDelete: jest.fn(),
    friendUpdate: jest.fn(),
    messageToast: jest.fn(),
    editStart: jest.fn(),
    editEnd: jest.fn(),
    ...specificProps,
  };
  const parent = shallow(<FriendDetailsHeader {...props} />);
  const level = [];
  level[0] = parent;
  level[1] = parent.dive();
  // level[2] = level[1].dive();
  const component = level[1];
  return {
    component,
    parent,
    level,
    props,
  };
}

describe('Rendering', () => {
  it('should render, when not editing', () => {
    const { component } = setup();
    expect(component).toMatchSnapshot();
  });

  it('should render, when not editing', () => {
    const { component } = setup({ isEditing: false, editedObject: {} });
    expect(component).toMatchSnapshot();
  });
});

/* ========= Functions ========= */

describe('Functions, when editing', () => {
  const { component, props } = setup();
  const headerModular = component.find(HeaderModular);

  describe("when calling cancelButton's action", () => {
    headerModular.prop('cancelButton').action();

    it('navigates', () => {
      expect(props.navigation.goBack).toBeCalled();
    });

    it('ends edit', () => {
      expect(props.editEnd).toBeCalled();
    });
  });

  describe('when calling first actionButton action', () => {
    headerModular.prop('actionButtons')[0].action();

    it('starts editing', () => {
      expect(props.editStart).toBeCalled();
    });
  });

  describe('when calling second actionButton action', () => {
    headerModular.prop('actionButtons')[1].action();

    it('updates friend', () => {
      expect(props.friendUpdate).toBeCalledWith(props.editedObject);
    });

    it('messages', () => {
      expect(props.messageToast).toBeCalled();
    });

    it('ends editing', () => {
      expect(props.editEnd).toBeCalled();
    });
  });

  describe('when calling third actionButton action', () => {
    headerModular.prop('actionButtons')[2].action();

    it('navigates', () => {
      expect(props.navigation.goBack).toBeCalled();
    });

    it('updates friend', () => {
      expect(props.friendDelete).toBeCalledWith(props.friendId);
    });

    it('messages', () => {
      expect(props.messageToast).toBeCalled();
    });
  });
});

/* ========= Functions ========= */

describe('Functions, when NOT editing', () => {
  const { component, props } = setup({ isEditing: false, editedObject: {} });
  const headerModular = component.find(HeaderModular);

  describe('when calling second actionButton action', () => {
    headerModular.prop('actionButtons')[1].action();

    it('doesnt updates friend', () => {
      expect(props.friendUpdate).not.toBeCalled();
    });

    it('doesnt messages', () => {
      expect(props.messageToast).not.toBeCalled();
    });

    it('ends editing', () => {
      expect(props.editEnd).toBeCalled();
    });
  });
});
