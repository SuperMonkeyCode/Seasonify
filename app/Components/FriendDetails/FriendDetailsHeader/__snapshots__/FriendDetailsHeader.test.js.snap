// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering should render, when not editing 1`] = `
<withHandlers(pure(HeaderModular))
  actionButtons={
    Array [
      Object {
        "action": [Function],
        "icon": "create",
        "visibleIf": false,
      },
      Object {
        "action": [Function],
        "hideByDefault": true,
        "icon": "checkmark",
        "visibleIf": true,
      },
      Object {
        "action": [Function],
        "icon": "trash",
      },
    ]
  }
  cancelButton={
    Object {
      "action": [Function],
      "icon": "arrow-back",
    }
  }
  title="Some friend"
/>
`;

exports[`Rendering should render, when not editing 2`] = `
<withHandlers(pure(HeaderModular))
  actionButtons={
    Array [
      Object {
        "action": [Function],
        "icon": "create",
        "visibleIf": true,
      },
      Object {
        "action": [Function],
        "hideByDefault": true,
        "icon": "checkmark",
        "visibleIf": false,
      },
      Object {
        "action": [Function],
        "icon": "trash",
      },
    ]
  }
  cancelButton={
    Object {
      "action": [Function],
      "icon": "arrow-back",
    }
  }
  title="Some friend"
/>
`;
